### DISCUSSION TITLE
How to add a callback when a DB table is updated?

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#9968] [Usage Questions] [answered]
I'd like to trigger an action in my application when the table is updated, but can't understand how to achieve this after going through the events documentation.

Changes to the table are done by an update() on the Core table.

I've tried things like:
```
def my_callback(*args):
    print("Update")
listen(MyORMClass.__table__, "after_update", my_callback)
```

But, I either can't find any valid values and get an AttributeError (as is the case when passing the table), or the callback seems to never be called.

**Comment by zzzeek** [comment, marked as answer]
> Changes to the table are done by an update() on the Core table.

The only events for Core are [after_execute](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.ConnectionEvents.after_execute) and [after_cursor_execute](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.ConnectionEvents.after_cursor_execute).   You'd look at the given statement /context to determine if it involves the table you are interested in.

**Reply by Dreamsorcerer** [reply]
OK, so first argument would be the engine then, and the callback would look like:
```
    if isinstance(clauseelement, sa.Update):
        if clauseelement.entity_description["table"] is MyORMClass.__table__:
```

**Reply by zzzeek** [reply]
that would work, more specifically you could say `clauseelement.table is mytable`.   `.table` should be a public attribute for `Update`

**Reply by zzzeek** [reply]
this works with anything including using `Connection.execute()`.  If you were executing these Core update objects with a `Session` then you have some other event hooks too.
